Send LSN range in getpage request

knizhnik opened the pull request:
## Problem

There are three kinds of get_page :
 * 1. Master compute: get the latest page not older than specified LSN (horizon=Lsn::MAX)
 * 2. RO replica: get the latest page not newer than current WAL position replica already applied (horizon=GetXLogReplayRecPtr(NULL))
 * 3. Snapshot: get latest page not new than specified LSN (horizon=request_lsn)

The problem itself is more detailed explained in Slack:
https://neondb.slack.com/archives/C036U0GRMRB/p1705932864991419

So I assume that two cases are clear:

Normal RW node: it always request latest version of there page because nobody except it can update pages.
Static RO node (snapshot): it always request version of the page not greater than snapshot LSN
In both cases we do no need range and just "latest" flag.
Most difficult is hot-standby replica. It accepts and apply WAL concurrently with PS. Replica can be ahead of PS or visa versa. This is why replica can not request latest version of the page, because if PS is ahead of replica, it will get "future" version of the page. But we also can not request version of the page with LSN not greater than specified (as in case of static replica). Because we are using "last written LSN" cache to estimate when page was last updated. If page was not updated for a long time, then requesting page worth this LSN may cause PS to retrieve version of the page which was already collected by GC.


So to handle case 3 we need top pas range of LSN: low boundary specifies estimated LSN of last page update (so that we do make PS to wait for applying most recent updates) and higher boundary specifies current apply position of replica, to prevent PS from sending too young pages.

With such range determining start LSN position for lookup in layer map becomes even simpler than now with "latest" flag - it is just max(low_bloundary, min(last_record_lsn, upper_boundary))

## Summary of changes

Replace `latest` flag in get page request with `horizon`

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

jcsp commented:
- I've added Christian as a reviewer: please can you add someone from the storage team as a review for PRs that change the pageserver.
- I haven't reviewed the code, but I notice that there aren't any tests in this PR

knizhnik commented:
> * I've added Christian as a reviewer: please can you add someone from the storage team as a review for PRs that change the pageserver.
> * I haven't reviewed the code, but I notice that there aren't any tests in this PR

It is protocol change affecting `get_page_at_lsn` - the main PS API. So this PR is affecting all tests and so all tests are checking it. But certainly this protocol change was done to address  "Key already removed by GC issue" in PR  #6357.
And there is test_replication_lag.py in this PR. @hlinnaka ask to separate protocol changes in separate PR. Here it is. Unfortunately this PR itself is not enough to solve the access beyond GC cutoff issue.  This PR is necessary but not sufficient. We need also prevent GC from deleting LSN beyond replicas apply position. And here we need changes in #6357 which propagate standby feedback through safekeeper and broker to PS.

problame reviewed (dismissed):
Please add a test case that demonstrates the problem / regress-tests your solution.

Also, you are making a breaking change to the protocol.

Please provide a plan to orchestrate the rollout.
A Slack channel (e..g, `#proj-page_service-lsn-range-requests` ) is probably most appropriate, as this touches multiple components.

knizhnik commented:
> Please add a test case that demonstrates the problem / regress-tests your solution.
> 
> Also, you are making a breaking change to the protocol.
> 
> Please provide a plan to orchestrate the rollout. A Slack channel (e..g, `#proj-page_service-lsn-range-requests` ) is probably most appropriate, as this touches multiple components.

I just wrote in the previous comment.
Problem is described in #6211 
It was expected to be fixed by #6357 and it contains test reproducing the problem.
This PR is separated from #6357.
It is not enough to fix the problem - only in conjunction with changes #6357 the problem will be fixed.
So I do not see much sense in moving test from @6357 to this PR.

Concerning rollout: changes in protocol are backward compatible - new server ids able to serve old clients. 
So we can deploy server and then sometimes later upgrade computes.

andreasscherbaum commented:
Discussion: add tests, also cover different protocols in tests

At the very least, describe how you tested it.

knizhnik commented:
I have added more comments explaining how backward compatibility of protocol is provided.
Also I have added `neon.protocol_version` GUC. It isn't sent now to page server, so ro provide backward compatibility is is necessary not to change format of existed commands. New protocol version should just add new commands.

`neon.protocol_version`  may be used at compute side to use specific protocol version.
It is currently used in `test_protocol_version.py` test  which tests backward compatibility. It can be also used if for some reasons we need new client to work with old page server.

andreasscherbaum commented:
Conflicts are resolved.

@problame @petuhovskiy Can you please have a look? Thanks.

knizhnik commented on pgxn/neon/pagestore_smgr.c:
```diff
@@ -110,6 +110,20 @@ static UnloggedBuildPhase unlogged_build_phase = UNLOGGED_BUILD_NOT_IN_PROGRESS;
+static XLogRecPtr
+neon_get_horizon(bool latest)
+{
+	return latest ? MAX_LSN : RecoveryInProgress() ? GetXLogReplayRecPtr(NULL) : InvalidXLogRecPtr; /* horizon=InvalidXlogRecPtr is replaced with request_lsn at PS */
```
No.
InvalidXLogrecPtr=0 is low boundary, MAX_LSN=~0 is upper boundary.
This is my comment from page_service.lsn:
```
        // Horizon = 0 (INVALID) is treated as LSN interval degenerated to point [lsn,lsn].
        // It as done mostly for convenience (because such get_page commands are widely used in tests) and
        // also seems to be logical: Lsn::MAX moves upper boundary of LSN interval till last_record_lsn and
        // Lsn(0) moves upper boundary to lower boundary.
```

knizhnik commented on pgxn/neon/pagestore_client.h:
```diff
@@ -79,7 +86,7 @@ typedef enum {
 {
 	NeonMessageTag tag;
-	bool		latest;			/* if true, request latest page version */
+	XLogRecPtr	horizon;		/* uppe boundary for page LSN */
```
Done

knizhnik commented on test_runner/regress/test_protocol_version.py:
```diff
@@ -0,0 +1,9 @@
+    endpoint = env.endpoints.create_start("main", config_lines=["neon.protocol_version=1"])
+    cur = endpoint.connect().cursor()
+    cur.execute("show neon.protocol_version")
+    assert cur.fetchone() == ("1",)
```
Right now we are not including catalog in basebackup: catalog pages are loaded on demand from PS as pages of any other relations.

knizhnik commented on libs/pageserver_api/src/models.rs:
```diff
@@ -948,8 +948,8 @@ impl PagestreamFeMessage {
-                bytes.put_u8(0);
-                bytes.put_u8(u8::from(req.latest));
+                bytes.put_u8(10);
+                bytes.put_u64(req.horizon.0);
```
Sorry, I do not understand the question.
What do you mean by old-style IO?

Yes, we have new format of commands (replacing boolean `latest` flag with uint64 `horizon`).
This is why command tag is changed. Page server recognise both old and new commands.

But in the models.rs we serialize only new commands.

knizhnik commented:
> Please add a test case that demonstrates the problem / regress-tests your solution.
> 
> Also, you are making a breaking change to the protocol.
> 
> Please provide a plan to orchestrate the rollout. A Slack channel (e..g, `#proj-page_service-lsn-range-requests` ) is probably most appropriate, as this touches multiple components.

New page server is able to server old and new commands. So backward compatibility is supported.
For compute I have added `neon.protocol_version` GUC which can be used to make neon extension to choose old or new format. We can first deploy compute with `neon.protocol_version=1`.
In this case client will send old commands.

We can rollback deploy of PS and still running clients will be able to work with old server.

Then at next deploy, we can switch protocol version to 2.

Right now default value for `neon.protocol_version` GUC is 2. I can change it to 1 or we can explicitly specify it in compute spec provided by control plane. What is better?

arssher reviewed (approved):
Good idea with compatible protocol upgrade, thanks!

I'd merge it with default proto version 1 and keep on staging for a day, and then change default to 2. Or keep it default 2 from the start, but spin up some replicas before merge and test them.

I wonder how easy would be to have a test specifically for the horizon & standby issue, exposing errors on standby when page which hadn't changes for a long time is requested. https://github.com/neondatabase/neon/pull/7368/ has test_replication_lag which covers this, but it needs both quite different issues to be fixed. Something like 1) create some page at some LSN 2) create heavy workload which moves this LSN beyond the horizon, but doesn't touch this page 3) wait for standby to catch up 3) make it request the page.

I'm quite confident with merging as this though because that test works and should be committed soon, primary has good test coverage and standby currently malfunctions anyway, but still having more isolated test would be great.

VladLazar commented on libs/pageserver_api/src/models.rs:
```diff
@@ -841,36 +841,36 @@ impl TryFrom<u8> for PagestreamBeMessageTag {
 #[derive(Debug, PartialEq, Eq)]
 pub struct PagestreamExistsRequest {
-    pub latest: bool,
+    pub horizon: Lsn,
```
I think the semantics would be clearer if we used a stronger type for the horizon and didn't rely on the `Lsn::MAX` and `Lsn::INVALID` values or replaced `horizon` and `lsn` with a `Range<Lsn>`.

Something like:

```
enum LsnHorizon {
  None,
  At(Lsn),
  Infinite
}
```

VladLazar replied:
While writing this comment I realised that I don't get why we need the `Lsn::INVALID` value for horizon. `horizon == Lsn::INVALID` should be equivalent to `latest=false`, so could we set horizon to the request lsn in these cases instead?

I think Arseny had a similar comment.

knizhnik replied:
Well, introducing new typer requires too much refactoring and frankly speaking I do not some significant benefits of presence of such type. 
Also it will not allow to use very simple and expressive formula which I really like:
```
let effective_lsn = Lsn::max(lsn, Lsn::min(request_horizon, last_record_lsn));
```

Concerning second question: yes it is not srtictly needed to pass Lsn::INVALID. It is just more convenient. Both in page server tests and in neon extension where decision whether to use latest version or not usually taken in the context where request lsn is not known.

hlinnaka replied:
> Well, introducing new typer requires too much refactoring and frankly speaking I do not some significant benefits of presence of such type. Also it will not allow to use very simple and expressive formula which I really like:
> 
> ```
> let effective_lsn = Lsn::max(lsn, Lsn::min(request_horizon, last_record_lsn));
> ```

That's very unclear to me. What is `request_horizon` ? Is somehow related to server-side GC horizon perhaps? Reading this, I keep getting confused what exactly is horizon, what the special LSN values mean, etc.

VladLazar commented on test_runner/regress/test_protocol_version.py:
```diff
@@ -0,0 +1,9 @@
+from fixtures.neon_fixtures import NeonEnv
+
+
+def test_protocol_version(neon_simple_env: NeonEnv):
```
Extending this test to restart the compute and do a few reads with the new protocol version would give me some extra peace of mind.

knizhnik replied:
Please notice that to execute even simplest query as `select 1` or `show neon.protocol_version` Postgres actually has to load a lot of pages from system catalog.

VladLazar replied:
Makes sense, but I think it would still make sense to restart with the new protocol version and execute `show neon.protocol_version`.

VladLazar commented on pgxn/neon/pagestore_client.h:
```diff
@@ -28,10 +28,17 @@
 {
 	/* pagestore_client -> pagestore */
-	T_NeonExistsRequest = 0,
+	T_NeonExistsRequest = 10, /* new protocol message tags start from 10 */
```
Wasn't obvious to me that this field encodes the protocol version. It would be nicer to somehow include `NEON_PROTOCOL` version in the right hand side of the assignment and make it clearer.

knizhnik replied:
I wrote the comment explaining that we do not have protocol version now and it is not possible to introduce it without breaking backward compatibility. This is why the requirement is that any changes of protocol require introducing of new command tags, preserving format of old commands unchanged.

I agree with you that explicit passing of protocol version will be better, but relying on commands tag also seems to be acceptable. At least I hope we will not change this protocol too frequently...

kelvich replied:
Now the whole compute<>pageserver communications start with `pagestream` keyword. We can start it with e.g. `pagestream_v2` to make new protocol version to be explicit.

knizhnik replied:
It is actually done in @hlinnaka 's version of this PR: #7377

VladLazar commented on libs/pageserver_api/src/models.rs:
```diff
@@ -979,11 +979,32 @@ impl PagestreamFeMessage {
+        //
+        let horizon = if msg_tag >= 10 {
+            // new protocol
+            msg_tag -= 10; // commands tags in new protocol starts with 10
```
Why do we need to mutate `msg_tag` here?

arssher replied:
It's a hack to workaround missing protocol version field. Instead msg_tag is abused, shifting all tags on 10 (we have only a few of them, so no valid values >= 10). To actually parse we need to subtract 10 back then.

hlinnaka replied:
I think it would be more clear to add new commands with new tags and add handlers for them, as if they're completely new commands. After we don't need the backwards compatibility anymore, remove the old ones, leaving their tags unused.

knizhnik replied:
Unfortunately there are now no request tags in Rust at all!
There is enum ` PagestreamFeMessage`:
```
pub enum PagestreamFeMessage {
    Exists(PagestreamExistsRequest),
    Nblocks(PagestreamNblocksRequest),
    GetPage(PagestreamGetPageRequest),
    DbSize(PagestreamDbSizeRequest),
    GetSlruSegment(PagestreamGetSlruSegmentRequest),
}
```
But it is not somehow mapped to numbers. Number are hardcoded in `PagestreamFeMessage::parse`:
```
        match msg_tag {
            0 => Ok(PagestreamFeMessage::Exists(PagestreamExistsRequest {
```
and  `PagestreamFeMessage::serialize`:
```
   pub fn serialize(&self) -> Bytes {
        let mut bytes = BytesMut::new();

        match self {
            Self::Exists(req) => {
                bytes.put_u8(0);
```
 So it can be certainly refactored, but should it new part of this PR?

koivunej replied:
We should use serde here, but it doesn't need to be done on this PR. There are already comments for using serde. What exists can be turned into bincode quite easily.

Looking at the tests, we don't have image-based testing; we only have roundtrip parsing. That is insufficient to ensure backwards-compatible changes, so these will need to be fixed at the same time.

VladLazar reviewed (commented):
I think this is correct.

knizhnik commented on pageserver/src/pgdatadir_mapping.rs:
```diff
@@ -270,7 +264,7 @@ impl Timeline {
         tag: RelTag,
         version: Version<'_>,
-        _latest: bool,
+        _horizon: Lsn,
```
This is why I preserve it. Just for symmetry;)

knizhnik commented:
> I wonder how easy would be to have a test specifically for the horizon & standby issue, exposing errors on standby when page which hadn't changes for a long time is requested. https://github.com/neondatabase/neon/pull/7368 has test_replication_lag which covers this, but it needs both...

In principle we have `get_raw_page_at_lsn` in neon_test_utils extension to request page at there particular LSN.
But still to reproduce error we need PS to advance gc_cutoff beyond this point. So I do not see any principle difference here from what  test_replication_lag.py does

knizhnik commented on pageserver/src/page_service.rs:
```diff
@@ -849,67 +849,47 @@ impl PageServerHandler {
-            }
+        let last_record_lsn = timeline.get_last_record_lsn();
+        // Horizon = 0 (INVALID) is treated as LSN interval degenerated to point [lsn,lsn].
+        // It as done mostly for convenience (because such get_page commands are widely used in tests) and
```
> Why it is hard for them to use horizon=lsn?

I don't know best terms, but compute need to specify:
1. "desired" LSN - which LSN compute expects to be acceptable
2. upper boundary LSN - PS should not send page with greater LSN to preserver consistency

If `latest=true` (primary node), then upper boundary is always +inf: nobody except this node can produce more recent version of the page.
In case of replica it is not true: replica can lag from primary node and PS and should not receive pages newer than its `last_replay_lsn`. But it is not good always to request pages at  `last_replay_lsn` because replica can be **ahead** PS and so it has to wait until PS caught up (while for this partitucular page it is not needed).

So actually we need to handle just three cases:
1. [page_last_written_lsn, +inf] - primary node
2. [page_last_written_lsn, last_replay_lsn] - hot-standby replica (receiving WAL from primary)
3. [snapshot_lsn, shapshot_lsn] - static RO replica (not receiving WAL fro primary)

Case [0, lsn] is not actually needed and added mostly for convenience as alias for [lsn,lsn]

knizhnik commented on pageserver/src/page_service.rs:
```diff
@@ -849,67 +849,47 @@ impl PageServerHandler {
                     ctx,
                 )
                 .await?;
+            // Since we waited for 'lsn' to arrive, that is now the last
```
I have not written this comment and just copy it from original sources.

koivunej replied:
Maybe it should have
```suggestion
            // Since we waited for 'effective_lsn' to arrive, that is now the last
```

knizhnik commented on pageserver/src/page_service.rs:
```diff
@@ -849,67 +849,47 @@ impl PageServerHandler {
     /// the LSN that should be used to look up the page versions.
     async fn wait_or_get_last_lsn(
```
I added this comment. Sorry if it is not quite clear but I did my nest to explain why do we need range.

hlinnaka replied:
I find these 'horizon' and 'lsn' to be an unnatural way of thinking about this. My mental model is:

- the Compute always wants to see the data at one particular LSN. For a standby, it's the last-replayed LSN. For the primary, it's the last inserted LSN . Let's call this the "request LSN".
- In addition to the request LSN, the compute may send an additional hint to the pageserver. The hint tells the pageserver, "I know that the page hasn't been modified between the hint LSN, and the request LSN". If the compute doesn't know of any better, it can just use the request LSN for both.

I propose we call the fields in the Request structs, and elsewhere:
- simply `lsn` for the request LSN.
- `not_modified_since` for that hint.

arssher replied:
> My mental model is:

Yeah, that's much clearer.

hlinnaka commented on pageserver/src/pgdatadir_mapping.rs:
```diff
@@ -252,14 +252,8 @@ impl Timeline {
-            // latest=true, then it can not cause cache corruption, because with latest=true
-            // pageserver choose max(request_lsn, last_written_lsn) and so cached value will be
-            // associated with most recent value of LSN.
+        if horizon == Lsn::MAX {
+            // Update relation size cache only if latest version is requested.
             self.update_cached_rel_size(tag, version.get_lsn(), nblocks);
```
The old comment for why this is safe is still valid, right?

hlinnaka replied:
How significant is it to update the cache on reads anyway? I'd be inclined to just get rid of this, unless it makes a noticeable difference. I'm not 100% convinced the that reasoning is correct, and at least it depends on another assumption that hasn't been mentioned: it assumes that we haven't evicted any entries from the cache, since the WAL ingestion started. That currently holds, but at least we should document that somewhere.

Also, we should add a comment about that assumption about choosing max(request_lsn, last_written_lsn) also in wait_or_get_last_lsn, where we do that.  (BTW, shouldn't it be last_record_lsn rather than last_written_lsn?). We might actually want to change that, so that wait_or_get_last_lsn would use an older LSN if it can, because that might be more performant. Using an older LSN when digging through the layers could allow skipping the in-memory layer and recent L0 layers altogether, for example.

hlinnaka replied:
Here's a PR to refactor that part, so that we can eliminate those assumptions: https://github.com/neondatabase/neon/pull/7376. It's pretty orthogonal to this work, but would be nice to tighten that up.

koivunej commented on libs/pageserver_api/src/models.rs:
```diff
@@ -979,11 +979,32 @@ impl PagestreamFeMessage {
+        //
+        // Old version of protocol use commands with tags started with 0 and containing `latest` flag.
+        // New version of protocol shift command tags by 10 and pass LSN range instead of `latest` flag.
+        // Server should be able to handle both protocol version. As far as we are not passing no=w,
+        // protocol version from client to server, we make a decision based on tag range.
+        // So this code actually provides backward compatibility.
```
```suggestion
        // Server should be able to handle both protocol versions. As far as we are not passing now
        // protocol version from client to server, we decide based on tag range.
        // So, this code provides backward compatibility.
```

I was first thinking the `no=w` was some assignment which I couldn't figure out, but perhaps it was just a typo.

arssher commented:
superseded by
https://github.com/neondatabase/neon/pull/7377
